src/useMultiRootEditor.tsx

```diff
@@ -0,0 +1,508 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import React, { useState, useEffect, useRef, type Dispatch, type SetStateAction, useContext } from 'react';
+
+import type { EditorConfig } from '@ckeditor/ckeditor5-core';
+import type { DocumentChangeEvent, Writer, RootElement } from '@ckeditor/ckeditor5-engine';
+
+import { ContextWatchdog, EditorWatchdog } from '@ckeditor/ckeditor5-watchdog';
+import type { WatchdogConfig } from '@ckeditor/ckeditor5-watchdog/src/watchdog';
+
+import type { AddRootEvent, DetachRootEvent } from '@ckeditor/ckeditor5-editor-multi-root/src/multirooteditor';
+import type MultiRootEditor from '@ckeditor/ckeditor5-build-multi-root';
+import type EventInfo from '@ckeditor/ckeditor5-utils/src/eventinfo';
+
+import { ContextWatchdogContext } from './ckeditorcontext';
+import { EditorWatchdogAdapter } from './ckeditor';
+
+const REACT_INTEGRATION_READ_ONLY_LOCK_ID = 'Lock from React integration (@ckeditor/ckeditor5-react)';
+
+/* eslint-disable @typescript-eslint/no-use-before-define */
+const useMultiRootEditor = ( props: MultiRootHookProps ): MultiRootHookReturns => {
+	let watchdog: EditorWatchdog | EditorWatchdogAdapter<MultiRootEditor> | null = null;
+
+	const editorDestructionInProgress = useRef<Promise<void> | null>( null );
+
+	const context = useContext( ContextWatchdogContext );
+
+	// Current editor instance. It may change if the editor is re-initialized by the Watchdog after an error.
+	const [ editor, setEditor ] = useState<MultiRootEditor | null>( null );
+
+	// Current editor content. An object where each key is a root name and the value is the root content.
+	const [ content, setContent ] = useState<Record<string, string>>( props.content );
+
+	// Current roots attributes. An object where each key is a root name and the value is an object with root attributes.
+	const [ attributes, setAttributes ] = useState<Record<string, Record<string, unknown>>>( props.rootsAttributes || {} );
+
+	// Contains the JSX elements for each editor root.
+	const [ elements, setElements ] = useState<Array<JSX.Element>>( [] );
+
+	const shouldUpdateEditor = useRef<boolean>( true );
+
+	// The reference for the toolbar element.
+	const toolbarRef = useRef<HTMLDivElement>( null );
+
+	const toolbarElement = <div ref={ toolbarRef }></div>;
+
+	useEffect( () => {
+		const initEditor = async () => {
+			// When the component has been remounted it is crucial to wait for removing the old editor
+			// and cleaning the old state.
+			await editorDestructionInProgress.current;
+
+			if ( props.isLayoutReady !== false ) {
+				_initializeEditor();
+			}
+		};
+
+		initEditor();
```

Comment:
Why declare as separate function if it is used just once? Leftover after refactor?

Reply:
It is some kind of convention regarding creating async calls in `useEffect` hook. You can read the answer from the main React engineer: https://github.com/facebook/react/issues/14326#issuecomment-441680293